### ISSUE TITLE
Issue updating to 1.14.1: "PREV for xxxx is None"

### ISSUE THREAD

**Issue by drussel** [conan-io/conan#4898]
We are trying to update from 1.13.1 to 1.14.1. However, when we run `path/to/python.exe -m conans.conan install path/to/parent/of/conanfile.py version@user/channel -g cmake -if path/to/working/dir --build missing -s "compiler=Visual Studio" -s compiler.version=15 -o tbbmalloc=True -s build_type=Release` we get an error

```
Traceback (most recent call last):
  File "build\venv\lib\site-packages\conans\model\ref.py", line 111, in loads
    _, name, version, user, channel, revision, _ = ConanFileReference.sep_pattern.split(text)
ValueError: not enough values to unpack (expected 7, got 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "build\venv\lib\site-packages\conans\client\command.py", line 377, in install
    ref = ConanFileReference.loads(args.path_or_reference)
  File "build\venv\lib\site-packages\conans\model\ref.py", line 114, in loads
    "OpenCV/1.0.6@user/stable" % text)
conans.errors.ConanException: Wrong package recipe reference C:/Users/daniel/src/flow
Write something like OpenCV/1.0.6@user/stable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "build\venv\lib\site-packages\conans\client\command.py", line 1579, in run
    method(args[0][1:])
  File "build\venv\lib\site-packages\conans\client\command.py", line 390, in install
    install_folder=args.install_folder)
  File "build\venv\lib\site-packages\conans\client\conan_api.py", line 93, in wrapper
    return f(*args, **kwargs)
  File "build\venv\lib\site-packages\conans\client\conan_api.py", line 580, in install
    no_imports=no_imports)
  File "build\venv\lib\site-packages\conans\client\manager.py", line 80, in install
    installer.install(deps_graph, keep_build)
  File "build\venv\lib\site-packages\conans\client\installer.py", line 302, in install
    self._build(nodes_by_level, keep_build, root_node, graph_info)
  File "build\venv\lib\site-packages\conans\client\installer.py", line 324, in _build
    self._handle_node_cache(node, keep_build, processed_package_refs)
  File "build\venv\lib\site-packages\conans\client\installer.py", line 404, in _handle_node_cache
    assert node.prev, "PREV for %s is None" % str(pref)
AssertionError: PREV for Boost/1.68.0-16@velo3d/thirdparty:df18e90df1fedfd9db3e0852ce38a27c76c5a54b is None

ERROR: PREV for Boost/1.68.0-16@velo3d/thirdparty:df18e90df1fedfd9db3e0852ce38a27c76c5a54b is None
```

I'm at a bit of a loss as to what is going on. 

The `version@user/channel` in the command is something we were not previously using, but we get a complaint about a missing reference if it isn't there.

The boost package  should already be on the system, if that matters.

Thanks.

**Comment by drussel** [comment]
Conan is trying to build another package other than boost as part of this run.

**Comment by Johnnyxy** [comment]
I never seen that someone executes Conan this way. If you have installed Conan with PIP then you automatically get an executable ```conan.exe``` in your Python search path.
```
pip install -U conan
```
```
c:\tmp> conan install "path/to/parent/of/conanfile.py" "version@user/channel" -g cmake -if "path/to/working/dir" --build=missing -s compiler="Visual Studio" -s compiler.version=15 -o tbbmalloc=True -s build_type=Release
```

The complaint about a missing reference is correct according to the [docs](https://docs.conan.io/en/latest/reference/commands/consumer/install.html#conan-install). If you do not supply the name and/or version in the recipe you have to provide this information at Conan invocation.
- user/channel
- version@user/channel
- pkg/version@user/channel

**Comment by drussel** [comment]
We use a virtual environment so we can control the Conan version as part of our build process. This is only partly successful since we are caught between Conan changes that impact the ~/.conan files and those that impact recipes, but seems to work better than trying to trust the system Conan to be an appropriate version on all systems where we try to build (and that all projects agree on an appropriate version) or enforcing the Conan version as part of the build.

I'm not sure invoking it with `python -m conans.conan` does anything useful any more. It was a work around to deal with path/pythonpath variability prior to switching to the virtual env since getting a consistent behavior across various developers (windows) machines and Jenkins machines had proved tricky.

I'll see if calling `venv/[bin,Scripts]/conan` directly makes any difference.

**Comment by lasote** [comment]
I don't think to call conan that way makes any difference.
I'm trying to reproduce or figure out, but it is indeed a bug.
Thanks for reporting!

**Comment by lasote** [comment]
@drussel Could you please paste here the contents (if present) of the file in your conan cache at:
`~/.conan/data/Boost/1.68.0-16/velo3d/thirdparty/metadata.json`?
To confirm, you haven't removed that Boost package from the cache after trying to upgrade to `1.14.1`, right?

**Comment by david-sinuela-pix4d-dev** [comment]
We had a similar issue today on one of our mac nodes, with this error message:

```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/model/ref.py", line 111, in loads
    _, name, version, user, channel, revision, _ = ConanFileReference.sep_pattern.split(text)
ValueError: not enough values to unpack (expected 7, got 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/command.py", line 376, in install
    ref = ConanFileReference.loads(args.path_or_reference)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/model/ref.py", line 114, in loads
    "OpenCV/1.0.6@user/stable" % text)
conans.errors.ConanException: Wrong package recipe reference pix4d-magma
Write something like OpenCV/1.0.6@user/stable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/command.py", line 1578, in run
    method(args[0][1:])
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/command.py", line 389, in install
    install_folder=args.install_folder)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/conan_api.py", line 93, in wrapper
    return f(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/conan_api.py", line 580, in install
    no_imports=no_imports)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/manager.py", line 80, in install
    installer.install(deps_graph, keep_build)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/installer.py", line 302, in install
    self._build(nodes_by_level, keep_build, root_node, graph_info)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/installer.py", line 324, in _build
    self._handle_node_cache(node, keep_build, processed_package_refs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/installer.py", line 404, in _handle_node_cache
    assert node.prev, "PREV for %s is None" % str(pref)
AssertionError: PREV for proj/4.9.3-6@pix4d/stable:f8bda7f0751e4bc3beaa6c3b2eb02d455291c8a2 is None
```

The content of ~/.conan/data/OpenCV/3.4.2-8/pix4d/stable/metadata.json is:
```
{"recipe": {"revision": "0", "properties": {}}, "packages": {"142c4d9c725d85b48b038429a5bed8d5deeb386c": {"revision": "0", "recipe_revision": "0", "properties": {}}}}
```

and the content of ~/.conan/data/proj/4.9.3-6/pix4d/stable/metadata.json is:
```
{"recipe": {"revision": "0", "properties": {}}, "packages": {}}
```

**Comment by lasote** [comment]
@david-sinuela-pix4d-dev  Could you say if that package was also created with a previous conan version?

**Comment by lasote** [comment]
I've been able to reproduce the issue. You have the package `f8bda7f0751e4bc3beaa6c3b2eb02d455291c8a2` but the `metadata.json` file is not containing that package ID (in your case none of them).
We could patch it but the goal is to guess why it happened. 
Could you tell us how that package was generated? 
- Regular `conan create`?
- Maybe `conan export-pkg`?
- Executed `conan copy`?
... any other information you could consider relevant?

**Comment by david-sinuela-pix4d-dev** [comment]
@lasote The package was generated in a separate CI job that builds the dependencies and uploaded to a local conan server using conan package tools. Both the builder and the server were using conan 1.12.3 until a few days ago when we upgraded to version 1.14.0. Actually I just checked, the packages were produced in February, at that time the builder node had conan 1.9.1 and conan package tools 0.20.

The node that failed today consuming the packages had conan 1.12.3 before and has 1.14.0 now.

I don't know if I will be able to reproduce the issue again as the nodes will be recycled tonight.

**Comment by lasote** [comment]
Thanks for the info. I'm trying changing conan version and I'm not able to reproduce the issue. I think we might consider a package as "corrupted" or "invalid" if at the local cache it has no entry in the "packages" at the metadata.json. Will talk with the team.

**Comment by memsharded** [comment]
Hi @drussel @david-sinuela-pix4d-dev 

I have been investigating this, without much success. I have introduced a new assert in https://github.com/conan-io/conan/pull/4927, which might bring some new data. It is already in develop, so if you could please try it from the test package:

https://test.pypi.org/project/conan/1.15.0.dev1554729202/

```bash
$ pip install -i https://test.pypi.org/simple/ conan==1.15.0.dev1554729202
```

And run it against your project, that would help. Many thanks!

**Comment by drussel** [comment]
The metadata printed out is ` {"recipe":
{"revision": "0", "properties": {}}, "packages": {"b16eb78376abcccbbb95c0ccfd7fc1c01657972d": {"revision": "0", "recipe_
revision": "0", "properties": {}}, "3ec8b5bd6b1c2f5e8ca4facd1a389bef6866cd83": {"revision": "0", "recipe_revision": "0",
 "properties": {}}, "df18e90df1fedfd9db3e0852ce38a27c76c5a54b": {"revision": null, "recipe_revision": null, "properties"
: {}}}}`

**Comment by drussel** [comment]
Looking some more. I have 9 subdirs in my packages folder for that recipe, but it looks like only 3 in the metadata. The df18 one  (will the null revision) looks to be the main one that gets used. If I conan search for that recipe, I get 7 hits, one of which lists no settings out options (that one is not in the metadata). 

A lot of the initial package development and building occurs on the machine in question. And there have likely been failures and aborts of builds along the way.

The one with no options listed looks to be partial (no lib dir in the package).

**Comment by memsharded** [comment]
Thanks very much for the feedback. It seems that some ``null`` revisions slipped into the metadata even with the migrations. We are going to prepare and release a new patch version with a more thorough migration asap.

**Comment by memsharded** [comment]
Migration merged, will be released in 1.14.2

**Comment by david-sinuela-pix4d-dev** [comment]
Thanks for the quick fix. Just one note: we are still using 1.14.0 and since the nodes were recycled we have not seen the problem again.
